apps/keyboard/js/settings/layout_dictionary_downloader.js

```diff
@@ -0,0 +1,101 @@
+'use strict';
+
+(function(exports) {
+
+var LayoutDictionaryDownloader = function(imEngineId, dictFilename) {
+  this.imEngineId = imEngineId;
+  this.dictFilename = dictFilename;
+
+  this._xhr = null;
+};
+
+// Remote URLs to download the data. We require the remote server to have
+// CORS turned on since we are not asking for SystemXHR permission.
+LayoutDictionaryDownloader.prototype.REMOTE_URL =
+  'https://fxos.cdn.mozilla.net/dictionaries/%imEngineId/%ver/%dictFilename';
+
+// If the data schema changes in an incompatible way, we must bump the
+// version-to-download here.
+LayoutDictionaryDownloader.prototype.ENGINE_DATA_VERSIONS = Object.freeze({
+  'latin': 1
+});
+
+LayoutDictionaryDownloader.prototype.onprogress = null;
+
+LayoutDictionaryDownloader.prototype.load = function() {
+  if (this._xhr) {
+    throw new Error('LayoutDictionaryDownloader: Already downloading...');
+  }
+
+  var p = new Promise(function(resolve, reject) {
+    var url = this.REMOTE_URL
+      .replace('%imEngineId', this.imEngineId)
+      .replace('%ver', this.ENGINE_DATA_VERSIONS[this.imEngineId])
+      .replace('%dictFilename', this.dictFilename);
+
+    var xhr = this._xhr = new XMLHttpRequest();
+    xhr.open('GET', url, true);
+    xhr.responseType = 'arraybuffer';
+
+    xhr.onprogress = function(evt) {
+      if (typeof this.onprogress === 'function') {
+        this.onprogress(evt.loaded, evt.total);
+      }
+    }.bind(this);
+
+    xhr.onloadend = function() {
+      this._xhr = null;
+
+      var data = xhr.response;
+      if (xhr.status !== 404 && data && this._verifyData(data)) {
+        if (typeof this.onprogress === 'function') {
+          this.onprogress(data.byteLength, data.byteLength);
+        }
+
+        resolve(data);
+      } else {
+        reject(xhr.statusText);
```

Comment:
I need some clarification here: is that "data is probably valid when status isn't 404; if data isn't actually valid, report the status -- when the developer sees the status is 200 (or some OK status), then data on CDN is somehow invalid. if the developer sees the status isn't 200, then the CDN itself is broken."